src/benchmark/tpch_benchmark.cpp

```diff
@@ -0,0 +1,355 @@
+#include <chrono>
+#include <fstream>
+#include <iostream>
+#include <random>
+#include <string>
+
+#include "cxxopts.hpp"
+#include "json.hpp"
+#include "sql/sql_pipeline.hpp"
+#include "tpch/tpch_db_generator.hpp"
+#include "tpch/tpch_queries.hpp"
+#include "SQLParser.h"
+#include "SQLParserResult.h"
+
+/**
+ * This benchmark measures Hyrise's performance executing the TPC-H queries, it doesn't (yet) support running the TPC-H
+ * benchmark as it is specified.
+ * The benchmark offers a wide range of options (scale_factor, chunk_size, ...) but most notably it offers two modes:
+ * IndividualQueries and PermutedQuerySets. See docs on BenchmarkMode for details.
+ * The benchmark will stop issuing new queries if either enough iterations have taken place or enough time has passed.
+ *
+ * main() is mostly concerned with parsing the CLI options while TpchBenchmark::run() performs the actual benchmark
+ * logic.
+ */
+
+namespace opossum {
+
+/**
+ * IndividualQueries runs each query a number of times and then the next one
+ * PermutedQuerySets runs the queries as sets permuting their order after each run (this exercises caches)
+ */
+enum class BenchmarkMode { IndividualQueries, PermutedQuerySets };
+
+using Duration = std::chrono::high_resolution_clock::duration;
+using TimePoint = std::chrono::high_resolution_clock::time_point;
+
+// Used to config out()
+auto verbose_benchmark = false;
+
+/**
+ * @return std::cout if `verbose_benchmark` is true, otherwise returns a discarding stream
+ */
+std::ostream& out() {
+  if (verbose_benchmark) {
+    return std::cout;
+  }
+
+  // Create no-op stream that just swallows everything streamed into it
+  // See https://stackoverflow.com/a/11826666
+  class NullBuffer : public std::streambuf {
+   public:
+    int overflow(int c) { return c; }
+  };
+
+  static NullBuffer null_buffer;
+  static std::ostream null_stream(&null_buffer);
+
+  return null_stream;
+}
+
+struct QueryBenchmarkResult {
+  size_t num_iterations = 0;
+  Duration duration = Duration{};
+};
+
+using BenchmarkResults = std::unordered_map<size_t, QueryBenchmarkResult>;
```

Comment:
This could use a comment on what `size_t` stands for.

Reply:
introduced `using QueryId = size_t`

Reply:
* `QueryID` if we want to follow the project’s naming scheme ;)

Reply:
yep